### PR TITLE
Tweak Runtime API error handling

### PR DESCRIPTION
**Pull request by arichiardi**
This patch adds a bit more information to the assert message in case the
Runtime API does not return 200, in particular for /error calls. This way it's
easier to debug what really happened, before the actual runtime error was never
showing up in the CloudWatch logs.

### CODE REVIEW

File: runtime.cljs [hunk 1]
```diff
@@ -33,50 +37,76 @@
...
-                           :body   (-> (clj->js response)
-                                       js/JSON.stringify)}))))
-      (.then (fn [{:keys [status]
-                   :as response}]
-               (assert (= status 202) (str "Unexpected /invocation/response response:" (pr-str response)))))
-      (.catch (fn [err]
-                (post-error {:error err
-                             :context (:context @state)})))
-      (.then start)))
+  (let [state (atom nil)
```
Comment (grav) [inline runtime.cljs]: I'm not sure if let'ing the state is a good idea. Won't we leak an atom each time an invocation of `start` is done? 

I agree that `def`ing an atom is not the best way, but maybe a better way would be to pass the reference to the atom each time we call `start` and only make a new one on the initial invocation? 

In any case it should be another PR, since it's not related to error handling as far as I see.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,19 @@
+RUNTIME_PACKAGE = runtime.zip
```
Comment (grav) [inline Makefile]: Let's introduce the make file in another PR and just have this PR relate to error handling :-)

File: README.md [hunk 3]
```diff
@@ -44,6 +44,13 @@ You'll get an error in the end, but an executable will nevertheless be created i
 zip -j runtime.zip bootstrap /path/to/lumo runtime.cljs
```
Comment (grav) [inline README.md]: With a `make` PR we'll just nuke the docs about manual building.

File: runtime.cljs (hunk 1 above)
Comment (arichiardi) [reply runtime.cljs]: Yeah well, yes if you want to share state between warm invocations.

I might have misread it but the I thought the info in that atom was per-invocation only and strictly only necessary for the downstream promises.

Let me know the actual intent, of course I can change it back 😁

Comment (grav) [reply runtime.cljs]: Yes, it's correct, it's only intended for downstream promises, and having it in a `let` would seem to be the best choice, except I'm not sure what happens when we recur (call `start` again in a `then`) - does the old atom get garbage collected? If so, your solution is good. If not, we should either keep the old way with `def`'ing an atom, or simply pass on the atom in the `start` invocation.

In both cases, let's keep this refactor for a new PR, in order to keep one feature per PR :-)

**Comment by arichiardi** [comment]
Reverted back the refactoring but still found a bit of a problem so don't  merge this yet please!

File: runtime.cljs [hunk 4]
```diff
@@ -29,54 +34,81 @@
...
+                 ;; that case, we simply log the output and fail.
+                 (assert (successful? status) (str "Error from Runtime API\n"
+                                                   (-> {:url url
+                                                        :response res
+                                                        :error runtime-error}
+                                                       (clj->js)
+                                                       (js/JSON.stringify nil 2))))
                  res)))))
 
 (defonce state (atom nil))
```
Comment (arichiardi) [inline runtime.cljs]: @grav Reverted this to have global state, this should be good to go.

File: bootstrap [hunk 5]
```diff
@@ -1,8 +1,7 @@
...
 
-APP_PARTS=(${_HANDLER//\// })
-tmpfile=`mktemp`
-echo "(require '${APP_PARTS[0]})" | cat - /opt/runtime.cljs > $tmpfile
+bin_dir=${BIN_DIR:-/opt}
+app_parts=(${_HANDLER//\// })
 
-/opt/lumo $tmpfile
\ No newline at end of file
+"$bin_dir/lumo" -e "(require (quote ${app_parts[0]}))" -m runtime
```
Comment (grav) [inline bootstrap]: Hi @arichiardi 
I just tried your fork of the project (including using your musl-based build of `lumo`), and got it working after tweaking the `bootstrap` file a bit:
```
"$bin_dir/lumo" --classpath ".:${bin_dir}" -e "(require (quote ${app_parts[0]}))" -m runtime
```
If I don't include the $bin_dir explicitly, it cannot seem to find the `runtime.cljs`. Additionally, I have to (re-)add current dir to the classpath.

Can you test that change? I can make a PR to your fork if you want me to.

Comment (grav) [reply bootstrap]: @arichiardi Btw, I got S3 access working with your Lumo build. Great work! :-) I'll eventually add an example to this repo.

Comment (arichiardi) [reply bootstrap]: I would like myself to PR against your work if you want me too I guess, I don't mind either choice, let's pick one fork 😁😁

I will definitely try to add your change, I thought I had it working properly there but maybe not 🙃

Comment (arichiardi) [reply bootstrap]: Oh I see will fix things

Comment (arichiardi) [reply bootstrap]: Ok I changed this and rebase @grav. Sorry didn't have time for doing an end-to-end test but I think that if you did it we are probably good to proceed and fix things afterwards.

In any case your change makes definitely sense to me, so much that I don't know why it was not like that already before.

**Comment by grav** [comment]
I did an end-to-end test and it works nicely! Thanks for the work :-)
